# Biostat 257 HW2
## Caesar Z. Li
## UID: 704135662

## Question1 (5 pts) Recognize structure

From the construction of the model, we have:
$$
\boldsymbol{P}_{ij}=\begin{cases}
    \frac{p}{\sum_j a_{ij}}+\frac{1-p}{n}, & \text{if $r_{i}\ne0$ and $a_{ij}\ne0$}.\\
    \frac{1-p}{n}, & \text{if $r_{i}\ne0$ and $a_{ij}=0$}. \\
    \frac{1}{n}, & \text{if $r_{i}=0$}.
  \end{cases}
$$
Note that every element in this matrix has the term $\frac{1}{n}$ no matter what, and is otherwise a sparse matrix. Thus we can write it as the sum of a sparse matrix and a rank 1 matrix with every single term equal to $\frac{1}{n}$ (it is of rank one because all elements are equal):    
$$
\boldsymbol{P}=p\boldsymbol{B} \boldsymbol{A} + \frac{\boldsymbol{c}\boldsymbol{1}^T_n}{n},
$$     
where $\boldsymbol{B}$ is a diagonal matrix with $i$-th diagonal term equal to $\frac{1}{\sum_j a_{1j}}$ if $r_{i}\ne0$, or $0$ if $r_{i}=0$, i.e.,    
$$
\boldsymbol{B} = diag(\frac{1}{\sum_j a_{1j}}or 0 \;\; \frac{1}{\sum_j a_{2j}}or0 \;\; \dots \;\; \frac{1}{\sum_j a_{nj}}or0)
$$
<br/>
and $\boldsymbol{c}$ is an indicator vector indicating the rows that are not dangling nodes,
$$
\boldsymbol{c}_{ij}=\begin{cases}
    \frac{1-p}{n}, & \text{if $r_{i}\ne0$ and $a_{ij}\ne0$}.\\
    \frac{1}{n}, & \text{if $r_{i}=0$}.
  \end{cases}
$$

## Q3. (10 pts) Explore data

In [1]:
using MatrixDepot

md = mdopen("SNAP/web-Google")
# display documentation for the SNAP/web-Google data
mdinfo(md)

include group.jl for user defined matrix generators
verify download of index files...
used remote site is https://sparse.tamu.edu/?per_page=All
populating internal database...


# SNAP/web-Google

###### MatrixMarket matrix coordinate pattern general

---

  * UF Sparse Matrix Collection, Tim Davis
  * http://www.cise.ufl.edu/research/sparse/matrices/SNAP/web-Google
  * name: SNAP/web-Google
  * [Web graph from Google]
  * id: 2301
  * date: 2002
  * author: Google
  * ed: J. Leskovec
  * fields: name title A id date author ed kind notes
  * kind: directed graph

---

  * notes:
  * Networks from SNAP (Stanford Network Analysis Platform) Network Data Sets,
  * Jure Leskovec http://snap.stanford.edu/data/index.html
  * email jure at cs.stanford.edu
  * 
  * Google web graph
  * 
  * Dataset information
  * 
  * Nodes represent web pages and directed edges represent hyperlinks between them.
  * The data was released in 2002 by Google as a part of Google Programming
  * Contest.
  * 
  * Dataset statistics
  * Nodes   875713
  * Edges   5105039
  * Nodes in largest WCC    855802 (0.977)
  * Edges in largest WCC    5066842 (0.993)
  * Nodes in largest SCC    434818 (0.497)
  * Edges in largest SCC    3419124 (0.670)
  * Average clustering coefficient  0.6047
  * Number of triangles     13391903
  * Fraction of closed triangles    0.05523
  * Diameter (longest shortest path)    22
  * 90-percentile effective diameter    8.1
  * 
  * Source (citation)
  * 
  * J. Leskovec, K. Lang, A. Dasgupta, M. Mahoney. Community Structure in Large
  * Networks: Natural Cluster Sizes and the Absence of Large Well-Defined Clusters.
  * arXiv.org:0810.1355, 2008.
  * 
  * Google programming contest, 2002
  * http://www.google.com/programming-contest/
  * 
  * Files
  * File    Description
  * web-Google.txt.gz   Webgraph from the Google programming contest, 2002

---

916428 916428 5105039


In [2]:
# connectivity matrix
MatrixDepot.addmetadata!(md.data)
A = md.A

916428×916428 SparseArrays.SparseMatrixCSC{Bool,Int64} with 5105039 stored entries:
  [11343 ,      1]  =  1
  [11928 ,      1]  =  1
  [15902 ,      1]  =  1
  [29547 ,      1]  =  1
  [30282 ,      1]  =  1
  [31301 ,      1]  =  1
  [38717 ,      1]  =  1
  [43930 ,      1]  =  1
  [46275 ,      1]  =  1
  [48193 ,      1]  =  1
  [50823 ,      1]  =  1
  [56911 ,      1]  =  1
  ⋮
  [608625, 916427]  =  1
  [618730, 916427]  =  1
  [622998, 916427]  =  1
  [673046, 916427]  =  1
  [716616, 916427]  =  1
  [720325, 916427]  =  1
  [772226, 916427]  =  1
  [785097, 916427]  =  1
  [788476, 916427]  =  1
  [822938, 916427]  =  1
  [833616, 916427]  =  1
  [417498, 916428]  =  1
  [843845, 916428]  =  1

In [3]:
# How much memory does A take? If converted to dense matrix (answered below)
@show Base.summarysize(A)

# number of web pages: # of columns/rows in A
@show size(A)
# number of edges (web links): simply use nnz function to count filled elements
@show nnz(A) 
# number of dangling nodes: first conduct row sums and then sum all that are equal to zero
@show sum(sum(A, dims = 2) .== 0)

Base.summarysize(A) = 53276943
size(A) = (916428, 916428)


UndefVarError: UndefVarError: nnz not defined

From above results, we can see that the memory usage was 53276943 bits, which is around 6.65962 Mbs, we have a total of 916428 pages, 5105039 web links, and 176974 dangling nodes.     
<br/>

Also, to answer the first question, since we know that the matrix is $916428\times916428$, and the type is `boolean`, the dense matrix would have taken 

$$
\frac{916428\times916428}{8 \; \text{bits/byte}} \approx 1.0498\times 10^{11} \; \text{bytes}
$$

or 104.98 GB, which is too large to be loaded into most computer's RAM.

In [4]:
using UnicodePlots

# histogram of in-degrees: histogram of column sums
degrees_in = vec(sum(A, dims = 1))
histogram(degrees_in, nbins = 30)

                    ┌                                        ┐ 
   [   0.0,  200.0) ┤▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 915169   
   [ 200.0,  400.0) ┤ 812                                      
   [ 400.0,  600.0) ┤ 203                                      
   [ 600.0,  800.0) ┤ 80                                       
   [ 800.0, 1000.0) ┤ 30                                       
   [1000.0, 1200.0) ┤ 19                                       
   [1200.0, 1400.0) ┤ 12                                       
   [1400.0, 1600.0) ┤ 6                                        
   [1600.0, 1800.0) ┤ 6                                        
   [1800.0, 2000.0) ┤ 9                                        
   [2000.0, 2200.0) ┤ 8                                        
   [2200.0, 2400.0) ┤ 4                                        
   [2400.0, 2600.0) ┤ 7                                        
   [2600.0, 2800.0) ┤ 11                                       
   [2800.0, 3000.0) ┤ 6                 

In [5]:
# list the top 20 pages with the largest in-degrees?
# Top 20 in-degree Page numbers are as shown below
@show sortperm(degrees_in, rev = true)[1:20]
# Top 20 in-degrees are as shown below
@show sort(degrees_in, rev = true)[1:20, :]

(sortperm(degrees_in, rev = true))[1:20] = [537040, 597622, 504141, 751385, 32164, 885606, 163076, 819224, 605857, 828964, 551830, 41910, 558792, 459075, 407611, 213433, 765335, 384667, 173977, 687326]
(sort(degrees_in, rev = true))[1:20, :] = [6326; 5354; 5271; 5182; 5097; 4847; 4731; 4620; 4550; 4484; 4220; 4219; 4206; 4187; 4180; 4084; 4015; 4010; 3988; 3956]


20×1 Array{Int64,2}:
 6326
 5354
 5271
 5182
 5097
 4847
 4731
 4620
 4550
 4484
 4220
 4219
 4206
 4187
 4180
 4084
 4015
 4010
 3988
 3956

In [6]:
# histogram of out-degrees: histogram of row sums

degrees_out = vec(sum(A, dims = 2))
histogram(degrees_out, nbins = 30)

                  ┌                                        ┐ 
   [  0.0,  20.0) ┤▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 891798   
   [ 20.0,  40.0) ┤▇ 22628                                   
   [ 40.0,  60.0) ┤ 1329                                     
   [ 60.0,  80.0) ┤ 371                                      
   [ 80.0, 100.0) ┤ 124                                      
   [100.0, 120.0) ┤ 86                                       
   [120.0, 140.0) ┤ 29                                       
   [140.0, 160.0) ┤ 15                                       
   [160.0, 180.0) ┤ 15                                       
   [180.0, 200.0) ┤ 8                                        
   [200.0, 220.0) ┤ 7                                        
   [220.0, 240.0) ┤ 3                                        
   [240.0, 260.0) ┤ 8                                        
   [260.0, 280.0) ┤ 3                                        
   [280.0, 300.0) ┤ 0                                        
   [300.

In [7]:
# which are the top 20 pages with the largest out-degrees?
# Again, first line outputs the page numbers to the top 20 out-degrees
@show sortperm(degrees_in, rev = true)[1:20]
# This second line outputs the actual top 20 in-degrees themselves
@show sort(degrees_out, rev = true)[1:20, :]

(sortperm(degrees_in, rev = true))[1:20] = [537040, 597622, 504141, 751385, 32164, 885606, 163076, 819224, 605857, 828964, 551830, 41910, 558792, 459075, 407611, 213433, 765335, 384667, 173977, 687326]
(sort(degrees_out, rev = true))[1:20, :] = [456; 372; 372; 330; 277; 268; 265; 258; 257; 256; 253; 248; 247; 245; 244; 231; 229; 225; 216; 213]


20×1 Array{Int64,2}:
 456
 372
 372
 330
 277
 268
 265
 258
 257
 256
 253
 248
 247
 245
 244
 231
 229
 225
 216
 213

In [8]:
# visualize the sparsity pattern of A or a submatrix of A say A[1:10000, 1:10000].

using UnicodePlots
spy(A[1:10000, 1:10000])

                      Sparsity Pattern
         ┌──────────────────────────────────────────┐    
       1 │⠀⠀⠀⠀⠀⠀⠀⠂⠀⠀⠀⠀⢀⢂⡂⠐⠀⠀⠀⠠⠀⠀⠀⠔⠄⠀⠈⢀⠀⠀⠉⠀⠀⠁⠀⠀⠐⠀⠄⠀⠐⠀│ > 0
         │⠀⠀⠀⠀⠀⠂⡂⠀⠄⠄⠈⠈⡔⠀⠀⠁⠂⠀⠀⠀⠀⠠⠁⠒⠁⠀⠄⠀⠐⠀⠈⠀⠀⡐⠀⠀⠠⠀⡀⠀⠉⠀│ < 0
         │⠀⠐⠠⠀⠀⠁⢐⠄⠀⠢⠠⠀⠀⡀⠈⠀⠋⠀⠀⠀⠀⠀⠂⡀⠀⠈⠄⠀⠄⠁⠈⠀⠠⠀⠰⠀⠀⠀⠀⢀⠆⠀│    
         │⠀⠈⠠⠂⠀⠈⠀⡠⠐⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⡀⠈⠄⠁⢀⡀⠐⠄⠄⠀⠀⢀⠀⠈⠂⢠⠀⠢⠐⠀⠀⠀⡀│    
         │⠀⠀⠐⡁⠀⡀⠀⠀⠠⠀⠁⠀⠀⠀⡄⠀⠈⠁⠀⠀⠀⠀⠠⠂⠀⠀⠀⠀⠀⠂⠪⠀⠀⠂⠀⠀⠀⠄⢀⠀⠁⠀│    
         │⠂⢀⠀⠁⠀⠀⡀⡉⠀⠀⠠⠂⠐⢀⠄⠁⠐⠀⠈⠀⠀⠀⠀⠀⡀⠐⠁⡠⠀⠂⠚⠀⠁⠀⠐⢀⠆⠀⠈⠄⢀⠂│    
         │⠀⠀⠀⠁⠀⡄⠀⠀⠀⠀⠒⠀⢄⠄⢂⠀⠂⠀⠀⠄⠀⢀⠀⠣⠀⢀⠀⠀⠀⠀⠠⡀⠀⠀⢠⠄⠀⡀⠀⠂⠂⠀│    
         │⢀⠈⠀⠀⠂⠀⠀⠁⢠⠈⠄⠁⠀⠀⡈⠈⠠⠀⢀⡦⠀⠀⠅⠐⠀⠨⠁⠨⠀⠄⠠⢀⠀⠐⢐⠀⠀⠉⠀⠀⠠⠉│    
         │⠀⠀⠨⠀⠋⠀⠄⠀⠂⠀⠈⠀⠀⠀⠈⠀⠀⠀⠀⡑⠀⢀⠈⠈⠀⠀⠀⠈⠀⠂⠐⠼⠀⠠⠀⠀⡈⠁⠀⠀⠡⠄│    
         │⠠⡁⠀⠀⠀⠀⠀⠀⠀⠀⠂⢀⠀⠀⠀⡄⠀⠀⠀⠤⠀⠀⠁⠀⡂⠀⠀⠀⠄⠀⢈⠀⢔⡊⣨⠈⠂⠀⠄⢀⠀⠀│    
         │⠀⠂⠀⠈⠀⠀⠀⠀⠀⠀⠈⠀⡀⠀⠂⠠⠐⠀⠄⠀⠀⠀⠀⠀⠀⠤⠀⠀⠀⠠⠘⠐⠀⠄⢘⠀⠀⢐⠄⠀⠁⠀│    
         │⢀⠄⢠⠄⡊⠂⠢⢀⠀⠀⠀⠀⠄⡀⡑⠀⡐⠀⠁⡂⠀⠀⠀⠰⠂⢠⠄⠈⠀⠈⠐⠀⠀⠀⠉⠀⠀⡀⠀⠀⠃⠀│    
         │⠀⠀⠅⠀⡁⠀⠀⠀⠀⠀⢁⢀⠀⡀⡀⡐⠀⠀⠈⠈⠔⠄⠀⠀⢀⠀⠀⠀⠀⠀⠀⠀⠀⠄⢉⠠⠁⠀⠁⠀⠀⠀│    
         │⠊⢈⠀⠁⠀⠀⠀⠅⠀⠂⠀⡀⠀⠀⠀⠀⡀⠀⠀⠐⠄⠁⠠⠀⠐⠈⠀⠀⠀⡀⢠⠀⠀⠀⢰⠀⠀⠀⢀⠀⠠⠂│    
         │⠈⠀⠀⠈⠄⠀⠀⠄⠐⠀⠠⠁⠁⠀⡂⠀⠠⠄⠌⠅⣀⠀⠀⢀⠀⠠⠀⠠⠠⠀⠲⢂⠀⠰⡄⠀⢀⠬⠈⠀⠀⠀│    
         │⠈⠀⠀⠄⠀⠀⡀⠈⠀⠀⠙⠀⠀⠀⢀⢀⠂⠀⠀⡄⢀⠀⠠

## Q4. (5 pts) Dense linear algebra

Answer: according to the LU decomposition lecture, to solve $\boldsymbol{Ax}=\boldsymbol{b}$ via LU, we need a total of 
$$
\frac{2}{3}n^3+2n^2 \approx 5.1310378\times10^{17} \; \text{flops}
$$
My macbook runs an Intel Quad-Core Intel Core i5, 2.3 GHz, the architecture of which processes 4 flops of Float64 per second. Let's be conservative and assume all four cores of my CPU are utilized (dividing by 4). Thus we'll need $\frac{5.1310378\times10^{17}}{4\times2.3\times10^{9}} \approx 5.577215\times10^{7}$ seconds to process, which is about $645.511$ days, or $1.769$ years to solve this equation.

As for memory allocation, if we are willing to overwrite matrix $\boldsymbol{A}$ with matrices $\boldsymbol{L}$ and $\boldsymbol{U}$, which will take about $1.0498\times10^{11}$ bytes (previous question), or 104.98 GB.

## Q5. Iterative solvers

### Step 1 (15 pts)

In [9]:
using BenchmarkTools, LinearAlgebra, SparseArrays, Revise

# a type for the matrix M = I - P^T in PageRank problem
struct PageRankImPt{TA <: Number, IA <: Integer, T <: AbstractFloat} <: AbstractMatrix{T}
    A         :: SparseMatrixCSC{TA, IA} # adjacency matrix
    telep     :: T
    # working arrays
    # TODO: whatever intermediate arrays you may want to pre-allocate

    B :: Vector{T}
    C :: Vector{T}
    storage_n1 :: Vector{T}
    storage_n2 :: Vector{T}
    #n :: IA
end

In [10]:
# constructor
function PageRankImPt(A::SparseMatrixCSC, telep::T) where T <: AbstractFloat
    n = size(A, 1)
    # TODO: initialize and pre-allocate arrays
    # Practically, just divide each row of A by its own row sum and plut (1-p)/n
    #B = A ./ sum(A, dims = 2) 
    #nn = nnz(A)
    B = Vector{Float64}(undef, n)
    C = similar(B)
   
    temp = vec(sum(A, dims = 2))
    for i in 1:n
        if temp[i] == 0
            B[i], C[i] = 0, 1 / n
        else 
            B[i], C[i] = telep / temp[i], (1 - telep) / n
        end
    end
    
    #BA = similar(A)
    #BA = (B * A) * telep
    
    #C .= C_temp .* telep / n
    #C_diag = similar(BA)
    #C_diag .= spdiagm(0 => C_temp) .* telep
    
    storage_n1 = similar(B)
    storage_n2 = similar(B)
    PageRankImPt(A, telep, B, C, storage_n1, storage_n2)
end

LinearAlgebra.issymmetric(::PageRankImPt) = false
Base.size(M::PageRankImPt) = size(M.A)
#Base.size(M::PageRankImPt) = M.n
# TODO: implement this function for evaluating M[i, j]
function Base.getindex(M::PageRankImPt, i, j)
    if i == j
        return 1 - M.A[j, i] * M.B[j] - M.C[j] 
    else 
        return - M.A[j, i] * M.B[j] - M.C[j] 
    end
end

# overwrite `out` by `(I - Pt) * v`
function LinearAlgebra.mul!(
        out :: Vector{T}, 
        M   :: PageRankImPt{<:Number, <:Integer, T}, 
        v   :: Vector{T}) where T <: AbstractFloat
    # TODO: implement mul!(out, M, v)
    # objective: compute A'Bv
    # compute Bv first and store in the first container
    M.storage_n1 .= M.B .* v
    # compute B(Av) and joggle to the second container
    mul!(M.storage_n2, transpose(M.A), M.storage_n1)
    # Compute 1(c'v) in efficient order 
    temp_sumv = dot(M.C, v)

    #sleep(1e-2) # wait 10 ms as if your code takes 1ms
    out .= v .- M.storage_n2 .- temp_sumv
end

# overwrite `out` by `(I - P) * v`
function LinearAlgebra.mul!(
        out :: Vector{T}, 
        Mt  :: Transpose{T, PageRankImPt{TA, IA, T}}, 
        v   :: Vector{T}) where {TA<:Number, IA<:Integer, T <: AbstractFloat}
    M = Mt.parent
    # TODO: implement mul!(out, transpose(M), v)
    # compute Av and store in first container
    mul!(M.storage_n1, M.A, v)
    # compute BAv and joggle to the second container
    M.storage_n1 .*= M.B
    # Compute c(1'v) in efficient order 
    temp_sumv = sum(v)
    M.storage_n2 .= M.C .* temp_sumv
    
    #sleep(1e-2) # wait 10 ms as if your code takes 1ms
    out .= v .- M.storage_n1 .- M.storage_n2
end

In [11]:
using DelimitedFiles

isfile("pgrksol.csv") || download("https://raw.githubusercontent.com/ucla-biostat-257-2020spring/ucla-biostat-257-2020spring.github.io/master/hw/hw3/pgrksol.csv",
                                  "pgrksol.csv")
xsol = vec(readdlm("pgrksol.csv"))

916428-element Array{Float64,1}:
 3.3783428216975054e-5
 2.0710155392568165e-6
 3.663065984832893e-6
 7.527510785028837e-7
 8.63328599674051e-7
 1.769418252415541e-6
 2.431230382883396e-7
 6.368417180141445e-7
 4.744973703681939e-7
 2.6895486110647536e-7
 3.18574314847409e-6
 7.375106374416742e-7
 2.431230382883396e-7
 ⋮
 1.1305006040148547e-6
 4.874825281822915e-6
 3.167946973112519e-6
 9.72688040308568e-7
 6.588614479285245e-7
 7.737011774300648e-7
 2.431230382883396e-7
 1.6219204214797293e-6
 3.912130060551738e-7
 2.431230382883396e-7
 7.296033831163157e-6
 6.330939996912478e-7

In [12]:
M = PageRankImPt(A, 0.85)
n = size(M, 1)

#@assert transpose(M) * ones(n) ≈ zeros(n)
@assert norm(transpose(M) * ones(n)) < 1e-12

In [13]:
@assert norm(M * xsol) < 1e-12

### Step 2 (20 pts)

In [14]:
M = PageRankImPt(A, 0.85)
n = size(M, 1)
v, out = ones(n), zeros(n)
bm_mv = @benchmark mul!($out, $M, $v) setup=(fill!(out, 0); fill!(v, 1))

BenchmarkTools.Trial: 
  memory estimate:  16 bytes
  allocs estimate:  1
  --------------
  minimum time:     24.743 ms (0.00% GC)
  median time:      30.359 ms (0.00% GC)
  mean time:        33.626 ms (0.00% GC)
  maximum time:     59.865 ms (0.00% GC)
  --------------
  samples:          145
  evals/sample:     1

In [15]:
bm_mtv = @benchmark mul!($out, $(transpose(M)), $v) setup=(fill!(out, 0); fill!(v, 1))

BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     25.377 ms (0.00% GC)
  median time:      25.834 ms (0.00% GC)
  mean time:        25.982 ms (0.00% GC)
  maximum time:     29.125 ms (0.00% GC)
  --------------
  samples:          186
  evals/sample:     1

In [16]:
clamp(10 - median(bm_mv).memory / 100, 0, 10) + 
clamp(10 - median(bm_mtv).memory / 100, 0, 10)

19.84

### Step 3 (20 pts)
Let's first try to solve the PageRank problem by the GMRES method for solving linear equations.

In [17]:
using KrylovKit

# normalize in-degrees to be the start point
x0   = vec(sum(A, dims = 1)) .+ 1.0
x0 ./= sum(x0)

# right hand side
b = zeros(n)

# warm up (compilation)
linsolve(M, b, x0, issymmetric = false, isposdef = false, maxiter = 1) 
# output is complex eigenvalue/eigenvector
(x_gmres, info), time_gmres, = @timed linsolve(M, b, x0, issymmetric = false, isposdef = false)

(([3.378342822183467e-5, 2.0710155392451414e-6, 3.6630659852312195e-6, 7.527510785590206e-7, 8.633285997147142e-7, 1.769418252734902e-6, 2.4312303829080185e-7, 6.368417180736775e-7, 4.7449737037419693e-7, 2.689548611091997e-7  …  3.167946973972633e-6, 9.72688041021552e-7, 6.588614478531544e-7, 7.737011774701954e-7, 2.4312303829080185e-7, 1.6219204214219846e-6, 3.9121300606249305e-7, 2.4312303829080185e-7, 7.29603383131283e-6, 6.33093999666783e-7], ConvergenceInfo: one converged value after 3 iterations and 72 applications of the linear map;
norms of residuals are given by (7.821170022857733e-13,).
), 6.0618673, 1010852361, 0.2062442, Base.GC_Diff(1010852361, 137, 0, 123826, 6, 0, 206244200, 5, 2))

In [18]:
@assert norm(x_gmres - xsol) < 1e-8

In [19]:
clamp(20 / time_gmres * 20, 0, 20)

20.0

### Step 4 (20 pts)
Let's first try to solve the PageRank problem by the Arnoldi method for solving eigen problems.

In [20]:
# warm up (compilation)
eigsolve(M, x0, 1, :SR, issymmetric = false, maxiter = 1)
# output is complex eigenvalue/eigenvector
(vals, vecs, info), time_arnoldi, = @timed eigsolve(M, x0, 1, :SR, issymmetric = false)

((Complex{Float64}[-8.347285846003266e-15 + 0.0im], Array{Complex{Float64},1}[[0.005635826953806779 + 0.0im, 0.00034549143807825587 + 0.0im, 0.0006110808494124571 + 0.0im, 0.00012557561626034094 + 0.0im, 0.00014402240532801998 + 0.0im, 0.0002951783050398202 + 0.0im, 4.0558328286775626e-5 + 0.0im, 0.00010623935784851125 + 0.0im, 7.915671116194682e-5 + 0.0im, 4.486765066724555e-5 + 0.0im  …  0.0005284839899796505 + 0.0im, 0.00016226599147987173 + 0.0im, 0.00010991273837608696 + 0.0im, 0.00012907055855750535 + 0.0im, 4.055832828677563e-5 + 0.0im, 0.0002705723874320158 + 0.0im, 6.526302748332824e-5 + 0.0im, 4.055832828677563e-5 + 0.0im, 0.0012171406601327078 + 0.0im, 0.00010561415510698254 + 0.0im]], ConvergenceInfo: one converged value after 7 iterations and 99 applications of the linear map;
norms of residuals are given by (9.575729268334711e-14,).
), 11.2189324, 1584015125, 0.1541198, Base.GC_Diff(1584015125, 213, 0, 139459, 50, 0, 154119800, 6, 0))

In [21]:
@assert abs(Real(vals[1])) < 1e-8

In [22]:
x_arnoldi   = abs.(Real.(vecs[1]))
x_arnoldi ./= sum(x_arnoldi)
@assert norm(x_arnoldi - xsol) < 1e-8

In [23]:
clamp(20 / time_arnoldi * 20, 0, 20)

20.0

## Q6. (5 pts) Results
List the top 20 pages you found and their corresponding PageRank score. Do they match the top 20 pages ranked according to in-degrees?

Answer: using the same sorting function from earlier questions to sort the solution vecto,

In [24]:
sortperm(vec(x_gmres), rev = true)[1:20]

20-element Array{Int64,1}:
 597622
  41910
 163076
 537040
 384667
 504141
 486981
 605857
  32164
 558792
 551830
 765335
 751385
 425771
 908352
 173977
   7315
 213433
 885606
 819224

In [25]:
# Then sort the solution from Arnoldi,
sortperm(vec(x_arnoldi), rev = true)[1:20]

20-element Array{Int64,1}:
 597622
  41910
 163076
 537040
 384667
 504141
 486981
 605857
  32164
 558792
 551830
 765335
 751385
 425771
 908352
 173977
   7315
 213433
 885606
 819224

Looks like the two solvers found the same solutions at least for the top 20 pages. good!

## Q7. Be proud of yourself

Answer: sounds good, will do.